service/resource/vaultpki/spec.go

```diff
@@ -0,0 +1,8 @@
+package vaultpki
+
+import vaultapi "github.com/hashicorp/vault/api"
+
+type VaultPKIState struct {
+	Backend       *vaultapi.MountOutput
+	CACertificate string
```

Comment:
It took me a lot of time to wrap my head around what is going on here. This `string` serves a `bool` function in this line:

```
	if vaultPKIStateToCreate.CACertificate != "" {
```

The `GetDesiredState` function doesn't work in this resource, and let's not try to hide that. I suggest changing this field to `CACertificateExists bool`. That way we can get rid of placeholders and it'll become obvious when the cert is actually created, and what is the function of this field.

Reply:
> This `string` serves a bool function in this line. 

No it does not. We need three states and not two. I started working with `BackendExists` and `CAExists` and ran into problems that either create, delete or update screws the reconciliation. The default would be `false` on which we then created them during `ProcessCreateState`. There is no update procedure implemented so the zero value is returned by `GetUpdateState`. `ProcessCreateState` got called again and saw `false` and tried to create the resources again, but they already existed and the reconciliation failed with "already exists" errors. Having explicit state which we actively set to something enables us to reflect the requirements for creation, deletion and updates. I know this sucks but at this point I do not know how to improve this. Some resources are apparently just different and we need to deal with their specialties somehow.

Reply:
I miss something in the flow. This is how I understand it.

First loop:

1. GetCreateState returns `CAExists=false`.
2. ProcessCreateState creates CA.
3. GetUpdateState returns zero object.
4. ProcessUpdate does nothing.
5. End of reconciliation loop.

Next loop:

1. GetCreateState returns `CAExists=true`.
2. ProcessCreateState does nothing.
3. GetUpdateState returns zero object.
4. ProcessUpdate does nothing.
5. End of reconciliation loop.

I don't understand how `ProcessCreateState` can be called with state created by `GetUpdateState`.

Reply:
It is this. 

First loop:

1. GetCreateState returns CAExists=false.
2. ProcessCreateState creates CA.
3. GetUpdateState returns zero object (CAExists=false).
4. ProcessUpdateState does nothing.
5. ProcessCreateState creates CA and fails with already exists.

Reply:
See https://github.com/giantswarm/operatorkit/blob/57dd6cec0cc811e2bc2e2c8a56ea40f163673514/framework/framework.go#L342-L398.

Reply:
I see. IMO this means `GetUpdateState` should be return correct `createState`.

Reply:
I don't like this.

Reply:
Relying on zero value of:

```
	if vaultPKIStateToCreate.CACertificate != "" {
```

is like relying on zero value of

```
	if vaultPKIStateToCreate.CACertificateNotExists {
```

which I don't like.

In that case zero value of `CACertificateNotExists` is `false` which means CA certificate exists.  I still think we use `string` as `bool` and `GetUpdateState` implementation is incorrect. As long as `ProcessUpdate` invokes `PorcessCreate` inside the framework, `GetUpdateState` should always return computed `createState` IMO.

Reply:
It is not boolean. We have three states because deletion handles state reversed as opposed to creation. The framework is not perfect but I also do not know how to improve it right now. There is no simple answer to this because there are a lot of complex cases that the framework should consider. I want to implement creation and deletion in case the resource requires creation and deletion based on its nature. Being forced to implement updates even though the resource does not require it only to make the framework work is just wrong to me. I also thought about resources that you provide only the creation and/or deletion interface to not even need to implement the empty update methods. This would then not work in this case, but it should. 

I like your efforts on doing the right thing but I do not want to get lost in details now. Feel free to propose changes to the framework to make the implementations easier. For now I would like to go with what we have to solve real problems and make our customers happy. This refactoring here eats too much time already. Lets iterate and do the right thing as soon as we know what the right thing is.